_source/_posts/2023-07-28-scim-workshop.md

```diff
@@ -0,0 +1,1144 @@
+
+```
+const prisma = new PrismaClient();
+
+// To funnel users into their designated orgs
+const ORG_ID = 1;
+interface IUserSchema {
+    schemas: string[];
+    userName?: string;
+    id?: string;
+    name?: {
+        givenName: string;
+        familyName: string;
+    };
+    emails?: { primary: boolean, value: string, type: string }[];
+    displayName?: string;
+    locale?: string;
+    meta?: {
+        resourceType: string;
+    }
+    externalId?: string;
+    groups?: [];
+    password?: string;
+    active?: boolean;
+    detail?: string;
+    status?: number;
+}
+
+const defaultUserSchema: IUserSchema = {
+    schemas: ['urn:ietf:params:scim:schemas:core:2.0:User'],
+    locale: 'en-US',
+    groups: [],
+    meta: {
+        resourceType: 'User'
+    }
+};
+```
+
+Finally, we are ready to get started! Let's work on our first CRUD endpoint. 
+
+### Create User Function - POST /Users - 
+
+Let's refer to the [SCIM spec](https://datatracker.ietf.org/doc/html/rfc7644#section-3.3) on creating a user. To create a user, the spec says that the client sends a POST request containing a "User" to the `/Users` endpoint. In response to the POST, the server signals a successful creation with an HTTP status code 201 (Created) and returns a representation of the user created. If the server determines that the creation of the requested user conflicts with existing users (e.g., a "User" resource with a duplicate "userName"), the server MUST return HTTP status code 409 (Conflict) with a "scimType" error code of "uniqueness," as per Section 3.12.
+
+Our `/Users` endpoint can fulfill those requirements with the following code: 
+
+```
+// Create User Function
+// POST /scim/v2/Users
+// RFC Notes on Creating Users: https://www.rfc-editor.org/rfc/rfc7644#section-3.3
+scimRoute.post('/Users', passport.authenticate('bearer'), async (req, res) => {
+    // console.log('POST: /Users');
+    // Format to SCIM standard
+    const newUser: IUserSchema = req.body;
+    console.log(req.body)
+    const { emails, password } = newUser;
+    const externalId = newUser.externalId;
+    const active = newUser.active;
+    const givenName = newUser.name?.givenName ?? 'NAME';
+    const familyName = newUser.name?.familyName ?? 'MISSING';
+    const displayName = `${givenName} ${familyName}`;
+    // Pull out primary email address
+    const emailPrimary = emails.find(email => email.primary === true);
+    const email = emailPrimary.value;
+    console.log('email: ', email);
+    // Set displayName to name
+    const name = displayName;
+    // Check if the User exists in the database
+    const duplicateUser = await prisma.user.findFirst({
+        select: {
+            id: true,
+            email: true,
+            name: true,
+        },
+        where: {
+            email,
```

Comment:
Explain why you are checking `email` here rather than `externalId`

Reply:
@isemona why ARE you checking email? my explanation generator is drawing a blank -- is a duplicate defined as one with the same email?